src/pytorch_lightning/plugins/io/async_plugin.py

```diff
@@ -0,0 +1,57 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from concurrent.futures import ThreadPoolExecutor
+from typing import Any, Optional
+
+from pytorch_lightning.plugins.io.checkpoint_plugin import CheckpointIO
+from pytorch_lightning.plugins.io.wrapper import _WrappingCheckpointIO
+
+
+class AsyncCheckpointIO(_WrappingCheckpointIO):
+    """``AsyncCheckpointIO`` enables saving the checkpoints asynchronously using the ``ThreadPoolExecutor``.
+
+    .. warning::
+
+        This is currently an experimental plugin/feature and API changes are to be expected.
+
+    Args:
+        checkpoint_io: A checkpoint IO plugin that is used as the basis for async checkpointing.
+        max_workers: Number of threads.
+    """
+
+    def __init__(self, checkpoint_io: Optional["CheckpointIO"] = None, max_workers: int = 1) -> None:
+        super().__init__(checkpoint_io)
+
+        self._executor = ThreadPoolExecutor(max_workers=max_workers)
+        self._error: Optional[BaseException] = None
+
+    def save_checkpoint(self, *args: Any, **kwargs: Any) -> None:
+        """Uses the ``ThreadPoolExecutor`` to save the checkpoints using the base ``checkpoint_io``."""
+
+        def _save_checkpoint(*args: Any, **kwargs: Any) -> None:
+            try:
+                assert self.checkpoint_io is not None
+                self.checkpoint_io.save_checkpoint(*args, **kwargs)
+            except BaseException as e:
+                self._error = e
+
+        self._executor.submit(_save_checkpoint, *args, **kwargs)
+
+        if self._error:
+            raise self._error
```

Comment:
A comment is appropriate here that this only raises the error if an error was raised between the previous time save_checkpoint was called and now, because executor.submit is not blocking. 

And given that, you also need to raise in teardown, because technically the error could be raised in the very last checkpoint you save. 

Also, a super nit edge case, but it's not handling different errors from multiple threads. I still strongly suggest to keep it at 1 and not expose num_threads to user.

Reply:
> Also, a super nit edge case, but it's not handling different errors from multiple threads. I still strongly suggest to keep it at 1 and not expose num_threads to user.

well then it's just any error from any thread. Anyway, the training is going to be interrupted.

Reply:
But please consider add the error raising to the teardown, and adding the comment I suggested.